FWCore/Framework/interface/CallbackExternalWork.h

```diff
@@ -0,0 +1,193 @@
+                }
+              }
+              if (excptr) {
+                Base::taskList().doneWaiting(excptr);
+                return;
+              }
+
+              Base::queue().push(
+                  *group, [this, holder = std::move(holder), serviceToken, record, eventSetupImpl]() mutable {
+                    Base::callingContext().setState(ESModuleCallingContext::State::kRunning);
+                    std::exception_ptr exceptPtr;
+                    try {
+                      convertException::wrap([this, &holder, &serviceToken, &record, &eventSetupImpl] {
+                        ESModuleCallingContext const& context = Base::callingContext();
+                        auto proxies = Base::getTokenIndices();
+                        if (Base::postMayGetProxies()) {
+                          proxies = &((*Base::postMayGetProxies()).front());
+                        }
+                        TRecord rec;
+                        edm::ESParentContext pc{&context};
+                        rec.setImpl(record, Base::transitionID(), proxies, eventSetupImpl, &pc);
+                        ServiceRegistry::Operate operate(serviceToken.lock());
+                        record->activityRegistry()->preESModuleAcquireSignal_.emit(record->key(), context);
+                        struct EndGuard {
+                          EndGuard(EventSetupRecordImpl const* iRecord, ESModuleCallingContext const& iContext)
+                              : record_{iRecord}, context_{iContext} {}
+                          ~EndGuard() {
+                            record_->activityRegistry()->postESModuleAcquireSignal_.emit(record_->key(), context_);
+                          }
+                          EventSetupRecordImpl const* record_;
+                          ESModuleCallingContext const& context_;
+                        };
+                        EndGuard guard(record, context);
+                        (*acquireFunction_)(rec, holder);
+                      });
+                    } catch (cms::Exception& iException) {
+                      iException.addContext("Running acquire");
+                      exceptPtr = std::current_exception();
+                    }
+                    holder.doneWaiting(exceptPtr);
+                  });
+            });
+      }
+
+      WaitingTask* makeExceptionHandlerTask(WaitingTask* produceTask, oneapi::tbb::task_group* group) {
+        return make_waiting_task([this, produceTask, group](std::exception_ptr const* iException) {
+          std::exception_ptr excptr;
+          if (iException) {
+            excptr = *iException;
+          }
+          WaitingTaskHolder holder(*group, produceTask);
+          if (excptr) {
+            try {
+              convertException::wrap([excptr]() { std::rethrow_exception(excptr); });
+            } catch (cms::Exception& exception) {
+              exception.addContext("Running acquire and external work");
```

Comment:
It is the first thing that can get the exception from the ExternalWork and maybe from acquire if it got far enough. It can convert to cms::Exception, add the usual context, and acquire specific context.